tools/clang/lib/CodeGen/CGCall.cpp

```diff
@@ -2368,6 +2368,48 @@ static llvm::StoreInst *findDominatingStoreToReturnValue(CodeGenFunction &CGF) {
   return store;
 }
 
+// If a function decl returns a matrix then performs a simple walthrough
+// in the function body and extracts the matrix packing info from the return stmt.
+static bool GetMatrixPackingInfo(Stmt *S, bool *isRowMajor) {
```

Comment:
That's a good point and the answer is no :(. E.g., in the below function, it is possible that we could use different packings for different return statements. However, let's say even if we would go for AST based solution, I am not sure which packing order should we bubble up to the return type? Also, I am wondering what would be the use case where user would use different packing under different return paths.

```
float2x2 GetMatrix(int i) 
{ 
 if(i > 0) 
 { 
   #pragma pack_matrix (row_major) 
   float2x2 mat = {1, 2, 3, 4}; 
   return mat; 
 } 
 else 
 { 
  #pragma pack_matrix (column_major) 
   float2x2 mat = {2, 3, 3, 4}; 
   return mat; 
 } 
}
```

Reply:
Also, may be, for the current solution, I could possibly gather packing info for all return statement/paths and assert if we encounter different packing under different path. That way we would know if hit that scenario in future.

Reply:
Is a return implemented through a store? I thought orientation only impacts loads and stores of matrices, and they are always the same in llvm registers (arrays of row vectors, I think?). So, either:
- If returns go through load/store, then we can make the storage location always row or column-major and make sure the return stores in that format and the user loads from that format.
- If returns do not go through load/store, then I don't see why we have a problem since both the caller and callee should use our convention for the ordering of matrices in LLVM registers.

Reply:
Yes, the return happens through a store followed by load. As an additional detail, in the initial code gen, these stores and loads are done by calling builtin hlsl functions for col-major and row-major (e.g., `dx.hl.matldst.rowStore`). These calls get lowered to regular llvm ld/st in a later pass. Anyways, I see your point that we could rather ensure that the load follows the **same** orientation. One way to achieve that is when we need to emit a load, we could track the user of alloca to get the matching store and use the same orientation?